File: 2022/day1/python/main.py
import argparse
from os import path


def part_one(input: str):
    # cells = input.split("\n\n")
    # sums = []
    # for cell in cells:
    #     s = sum([int(c) for c in cell.split("\n")])
    #     sums.append(s)

    # return max(sums)

    return max(
        [
            sum(y)
            for y in [
                map(lambda x: int(x), x)
                for x in [cell.split("\n") for cell in input.split("\n\n")]
            ]
        ]
    )


def part_two(input):
    cells = input.split("\n\n")
    sums = []
    for cell in cells:
        s = sum([int(c) for c in cell.split("\n")])
        sums.append(s)

    sums = sorted(sums)
    sums.reverse()
    return sum(sums[0:3])


parser = argparse.ArgumentParser(prog="aoc")
parser.add_argument("part", type=int)
parser.add_argument("--example", default=False)


def main():
    args = parser.parse_args()
    if args.part not in (1, 2):
        raise TypeError("part argument must be one of: 1 or 2")

    in_file_name = "example.txt" if args.example else "input.txt"
    in_file_path = path.abspath(__file__ + f"/../../data/{in_file_name}")
    input = ""
    with open(in_file_path) as f:
        input = f.read()

    result = ""
    if args.part == 1:
        result = part_one(input)
    elif args.part == 2:
        result = part_two(input)
    else:
        raise Exception("unreachable")

    print(result)


if __name__ == "__main__":
    main()

File: 2022/day1/rust/src/main.rs
use clap::Parser;
use std::fs;
use std::path::PathBuf;

#[derive(Parser)]
#[command()]
struct CLI {
    #[arg(value_parser = clap::value_parser!(u16).range(1..3))]
    part: u16,
    #[arg(short, long)]
    example: bool,
}

fn read_data_file(filename: &str) -> String {
    let in_file_path = PathBuf::from(format!("../data/{}", filename));
    let in_file = fs::canonicalize(&in_file_path).expect("could not canonicalize file");
    fs::read_to_string(in_file)
        .expect("could not read file")
        .to_string()
}

fn cell_sum(cell: &str) -> i64 {
    let nums = cell.split("\n");
    let parsed: Vec<i64> = nums.map(|num| num.parse::<i64>().unwrap()).collect();
    parsed.iter().sum()
}

fn top_n(input: Vec<i64>, n: usize) -> Vec<i64> {
    let mut sorted = input.clone();
    sorted.sort();
    sorted.reverse();
    sorted[0..n].into()
}

fn part_one(input: String) -> String {
    // actual readable version
    // let cells = input.split("\n\n");
    // let sums = cells.map(|cell| cell_sum(cell));
    // let answer = sums.max().unwrap();
    // answer.to_string()

    // cheeky one-liner version
    input
        .split("\n\n")
        .map(|cell| cell_sum(cell))
        .max()
        .unwrap()
        .to_string()
}

fn part_two(input: String) -> String {
    let cells = input.split("\n\n");
    let sums: Vec<i64> = cells.map(|cell| cell_sum(cell)).collect();
    let top_three = top_n(sums, 3);
    let answer: i64 = top_three.iter().sum();
    answer.to_string()
}

fn main() {
    let cli = CLI::parse();

    let mut in_file_name = "input.txt";
    if cli.example {
        in_file_name = "example.txt";
    }

    let input = read_data_file(in_file_name);

    if cli.part == 1 {
        println!("{}", part_one(input))
    } else {
        println!("{}", part_two(input))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_p1() {
        let input = read_data_file("example.txt");
        let want = "24000";
        let got = part_one(input);
        assert_eq!(want, got)
    }

    #[test]
    fn example_p2() {
        let input = read_data_file("example.txt");
        let want = "45000";
        let got = part_two(input);
        assert_eq!(want, got);
    }
}
